File: tests/rosetta/transpiler/FS/determine-if-a-string-is-collapsible.fs
// Generated 2025-07-31 00:10 +0700

exception Return

let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _substring (s:string) (start:int) (finish:int) =
    let len = String.length s
    let mutable st = if start < 0 then len + start else start
    let mutable en = if finish < 0 then len + finish else finish
    if st < 0 then st <- 0
    if st > len then st <- len
    if en > len then en <- len
    if st > en then st <- en
    s.Substring(st, en - st)

let rec collapse (s: string) =
    let mutable __ret : obj array = Unchecked.defaultof<obj array>
    let mutable s = s
    try
        let mutable i: int = 0
        let mutable prev: string = ""
        let mutable res: string = ""
        let mutable orig: int = String.length s
        while i < (String.length s) do
            let ch: string = _substring s i (i + 1)
            if ch <> prev then
                res <- res + ch
                prev <- ch
            i <- i + 1
        __ret <- [|box res; box orig; box String.length res|]
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        let strings: string array = [|""; "\"If I were two-faced, would I be wearing this one?\" --- Abraham Lincoln "; "..111111111111111111111111111111111111111111111111111111111111111777888"; "I never give 'em hell, I just tell the truth, and they think it's hell. "; "                                                   ---  Harry S Truman "; "The better the 4-wheel drive, the further you'll be from help when ya get stuck!"; "headmistressship"; "aardvark"; "😍😀🙌💃😍😍😍🙌"|]
        let mutable idx: int = 0
        while idx < (Seq.length strings) do
            let s: string = strings.[idx]
            let r: obj array = collapse s
            let cs: obj = r.[0]
            let olen: obj = r.[1]
            let clen: obj = r.[2]
            printfn "%s" (((("original : length = " + (string olen)) + ", string = «««") + s) + "»»»")
            printfn "%s" (((("collapsed: length = " + (string clen)) + ", string = «««") + (unbox<string> cs)) + "»»»\n")
            idx <- idx + 1
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()

File: tests/rosetta/transpiler/php/determine-if-a-string-is-collapsible.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
$__start_mem = memory_get_usage();
$__start = _now();
  function collapse($s) {
  $i = 0;
  $prev = '';
  $res = '';
  $orig = strlen($s);
  while ($i < strlen($s)) {
  $ch = substr($s, $i, $i + 1 - $i);
  if ($ch != $prev) {
  $res = $res . $ch;
  $prev = $ch;
}
  $i = $i + 1;
};
  return [$res, $orig, strlen($res)];
};
  function main() {
  $strings = ['', '"If I were two-faced, would I be wearing this one?" --- Abraham Lincoln ', '..111111111111111111111111111111111111111111111111111111111111111777888', 'I never give \'em hell, I just tell the truth, and they think it\'s hell. ', '                                                   ---  Harry S Truman ', 'The better the 4-wheel drive, the further you\'ll be from help when ya get stuck!', 'headmistressship', 'aardvark', '😍😀🙌💃😍😍😍🙌'];
  $idx = 0;
  while ($idx < count($strings)) {
  $s = $strings[$idx];
  $r = collapse($s);
  $cs = $r[0];
  $olen = $r[1];
  $clen = $r[2];
  echo rtrim('original : length = ' . _str($olen) . ', string = «««' . $s . '»»»'), PHP_EOL;
  echo rtrim('collapsed: length = ' . _str($clen) . ', string = «««' . $cs . '»»»
'), PHP_EOL;
  $idx = $idx + 1;
};
};
  main();
$__end = _now();
$__end_mem = memory_get_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
